### DISCUSSION TITLE
More flexible upload destination

### DISCUSSION THREAD

**Discussion by sylqiu** [rpldy/react-uploady#301] [Q&A]
I am trying to find how to upload with a service API such as firebase cloud storage instead of the destination URL.

Is it possible currently? If would be really nice to be able to configure the sender.

**Comment by yoavniran** [comment]
hi @sylqiu 

It will be great to understand your use case better.
What is it that you need to be more flexible exactly?

Uploady's sender works well for http end-points. There is a great ability for flexibility and configuration even at run-time and at the [batch](https://github.com/rpldy/react-uploady/tree/master/packages/ui/uploady#usebatchstartlistener-event-hook)/[request](https://github.com/rpldy/react-uploady/blob/master/guides/DynamicParameters.md) level.

If you need to pass the upload not to a http request but to another API (ex: 3rd party SDK) then this [discussion](https://github.com/rpldy/react-uploady/discussions/278) will be useful. It sounds like you're asking about the same thing.

Providing a custom [send](https://github.com/rpldy/react-uploady/blob/d7b232c38297320714ced62497300270a17d9201/packages/core/sender/types/index.d.ts#L31) method (using Uploady's [props](https://github.com/rpldy/react-uploady/tree/master/packages/ui/uploady#props)) is your best approach.

**Comment by sylqiu** [comment]
Amazing! Thanks for your pointer, I am able to achieve the purpose by something like the following:
```js
import { ref, uploadBytes, StorageReference, getStorage } from "firebase/storage";
import { BatchItem, FILE_STATES, UploadData } from "@rpldy/shared";
import { OnProgress, SendMethod, SendOptions, SendResult } from "@rpldy/sender";

const storage = getStorage()

const getRequestData = (items: BatchItem[]) => {
  if (items.length > 1) {
    throw new Error(
      `Firebase storage accepts only single file upload,
       while ${items.length} received.`
    );
  }
  if (items.length == 0) {
    throw new Error("No upload items found.");
  }
  console.log(items[0])
  return items[0].file;
};

const sendRequest = (
  storageRef: StorageReference,
  items: BatchItem[]
): Promise<UploadData> =>
  uploadBytes(storageRef, <Blob>(<unknown>getRequestData(items)))
    .then((snapshot) => {
      console.log("upload successful")
      return {
        status: 1,
        state: FILE_STATES.FINISHED,
        response: null,
      };
    })
    .catch((error) => {
      console.log("upload failed")
      return {
        status: 0,
        state: FILE_STATES.ERROR,
        response: error,
      };
    });

export const sender: SendMethod = (
    items: BatchItem[],
    url: string | undefined,
    options: SendOptions,
    onProgress?: OnProgress
  ): SendResult => {
    console.log(url)
    const storageRef = ref(storage, url);

    return {
      request: sendRequest(storageRef, items),
      abort: () => false,
      senderType: "FIREBASE_CLOUD_STORAGE",
    };
  };
```
The ```send``` function is passed as an argument to the ```Uploady``` provider, and the ```url``` above is coming from the ```destination``` argument.

**Reply by yoavniran** [reply]
@sylqiu Very nice!
